zio/parquetio/iterator.go

```diff
@@ -0,0 +1,726 @@
+}
+
+// plainReader decodes values encoded with the parquet PLAIN encoding
+// for all types other than BOOLEAN.  PLAIN encoding is described at:
+// https://github.com/apache/parquet-format/blob/master/Encodings.md#plain-plain--0
+// Handle Parquet PLAIN encoding type
+type plainReader struct {
+	buf []byte
+}
+
+func (r *plainReader) nextBoolean() bool {
+	// panic("implement plain boolean reader")
+	return false
+}
+
+func (r *plainReader) nextInt32() int32 {
+	if len(r.buf) < 4 {
+		panic("underflow in PLAIN INT32")
+	}
+
+	ret := binary.LittleEndian.Uint32(r.buf[:4])
+	r.buf = r.buf[4:]
+	return int32(ret)
+}
+
+func (r *plainReader) nextInt64() int64 {
+	if len(r.buf) < 8 {
+		panic("underflow in PLAIN INT64")
+	}
+
+	ret := binary.LittleEndian.Uint64(r.buf[:8])
+	r.buf = r.buf[8:]
+	return int64(ret)
+}
+
+func (r *plainReader) nextFloat() float64 {
+	if len(r.buf) < 4 {
+		panic("underflow in PLAIN FLOAT")
+	}
+	v := binary.LittleEndian.Uint32(r.buf[:4])
+	r.buf = r.buf[4:]
+	return float64(math.Float32frombits(v))
+}
+
+func (r *plainReader) nextDouble() float64 {
+	if len(r.buf) < 8 {
+		panic("underflow in PLAIN DOUBLE")
+	}
+	v := binary.LittleEndian.Uint64(r.buf[:8])
+	r.buf = r.buf[8:]
+	return math.Float64frombits(v)
+}
+
+func (r *plainReader) nextByteArray() []byte {
+	if len(r.buf) < 4 {
+		panic("underflow in PLAIN BYTE_ARRAY")
+	}
+	ln := binary.LittleEndian.Uint32(r.buf[:4])
+	total := int(4 + ln)
+	if len(r.buf) < total {
+		panic("underflow in PLAIN BYTE_ARRAY")
+	}
+	ret := r.buf[4:total]
+	r.buf = r.buf[total:]
+	return ret
+}
+
+// dictionaryInt64Reader decodes INT64-typed values encoded with the
+// parquet PLAIN_DICTIONARY encoding described at:
+// https://github.com/apache/parquet-format/blob/master/Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8
+type dictionaryInt64Reader struct {
+	dict        []int64
+	indexReader *hybridReader
+}
+
+func newDictionaryInt64Reader(buf []byte, dict []int64) *dictionaryInt64Reader {
+	width := int(buf[0])
+	reader := newHybridReader(buf[1:], width)
+	return &dictionaryInt64Reader{dict, reader}
+}
+
+func (r *dictionaryInt64Reader) nextBoolean() bool {
+	panic("cannot read BOOLEAN from INT64 dictionary reader")
+}
+
+func (r *dictionaryInt64Reader) nextInt32() int32 {
+	panic("cannot read INT32 from INT64 dictionary reader")
+}
+
+func (r *dictionaryInt64Reader) nextInt64() int64 {
+	i := int(r.indexReader.nextInt64())
+	if i > len(r.dict) {
+		panic(fmt.Sprintf("dictionary index too large (%d>%d)", i, len(r.dict)))
+	}
+	return r.dict[i]
+}
+
+func (r *dictionaryInt64Reader) nextFloat() float64 {
+	panic("cannot read FLOAT from INT64 dictionary reader")
+}
+
+func (r *dictionaryInt64Reader) nextDouble() float64 {
+	panic("cannot read DOUBLE from INT64 dictionary reader")
+}
+
+func (r *dictionaryInt64Reader) nextByteArray() []byte {
+	panic("cannot read BYTE_ARRAY from INT64 dictionary reader")
+}
```

Comment:
Instead of doing all this panics it might be nice to have a baseReader that these all inherit from. E.g:
```
type baseReader struct {
   name string
}

func (r *baseReader) nextFloat() float64 {
	panic(fmt.Sprintf("cannot read FLOAT from %s reader", r.name)
}
```